src/fallbacks.jl

```diff
@@ -0,0 +1,121 @@
+## generic `Tables.rows` and `Tables.columns` fallbacks
+## if a table provides Tables.rows or Tables.columns,
+## we'll provide a default implementation of the dual
+
+# generic row iteration of columns
+struct ColumnsRow{T}
+    columns::T # a `Columns` object
+    row::Int
+end
+
+Base.getproperty(c::ColumnsRow, ::Type{T}, col::Int, nm::Symbol) where {T} = getproperty(getfield(c, 1), T, col, nm)[getfield(c, 2)]
+Base.getproperty(c::ColumnsRow, nm::Symbol) = getproperty(getfield(c, 1), nm)[getfield(c, 2)]
+Base.propertynames(c::ColumnsRow) = propertynames(c.columns)
+
+struct RowIterator{T}
+    columns::T
+end
+Base.eltype(x::RowIterator{T}) where {T} = ColumnsRow{T}
+Base.length(x::RowIterator) = length(getproperty(x.columns, propertynames(x.columns)[1]))
+schema(x::RowIterator) = schema(x.columns)
+
+function Base.iterate(rows::RowIterator, st=1)
+    st > length(rows) && return nothing
+    return ColumnsRow(rows.columns, st), st + 1
+end
+
+function rows(x::T) where {T}
+    if columnaccess(T)
+        return RowIterator(columns(x))
+    else
+        throw(ArgumentError("no default `Tables.rows` implementation for type: $T"))
+    end
+end
+
+# build columns from rows
+"""
+    Tables.allocatecolumn(::Type{T}, len) => returns a column type (usually AbstractVector) w/ size to hold `len` elements
+    
+    Custom column types can override with an appropriate "scalar" element type that should dispatch to their column allocator.
+"""
+allocatecolumn(T, len) = Vector{T}(undef, len)
+
+@inline function allocatecolumns(::Schema{names, types}, len) where {names, types}
+    if @generated
+        vals = Tuple(:(allocatecolumn($(fieldtype(types, i)), len)) for i = 1:fieldcount(types))
+        return :(NamedTuple{names}(($(vals...),)))
+    else
+        return NamedTuple{names}(Tuple(allocatecolumn(fieldtype(types, i), len) for i = 1:fieldcount(types)))
+    end
+end
+
+haslength(x) = x === Base.HasLength() || x === Base.HasShape{1}()
```

Comment:
Is this actually needed? For example, `Base.IteratorSize(Matrix) isa Base.HasShape{2}`, yet you can call `length` and use linear indexing just like with a vector.

Reply:
We could expand the check to be `x === Base.HasLength() || x isa Base.HasShape`

Reply:
Well, would you look at [that](https://github.com/JuliaLang/julia/blob/3fbdc9bbc97e4e5b1d682883ad9d14be5d65c862/base/generator.jl#L98)